Loading web template from archive with "war" extension causes an error

### Questions

Do not use this issue tracker to ask questions, instead use one of these [channels](https://vertx.io/community/). Questions will likely be closed without notice.

### Version

4.3.8

### Context

I encountered an exception while running vertx that uses vertx-web
which specifically uses war file extension to start (not jar or zip)
java -jar file.war

When loading a web template using
```
FreeMarkerTemplateEngine.create(vertx).render(data, "/static/iframe.ftl", response -> {
```
vertx tries to load the template to cache
```
    private Object findTemplateSource(String path) throws IOException {
        final Object result = templateLoader.findTemplateSource(path);

```
which tries to extract the template file from archive but fails as expecting jar or zip archive extension.
```
      String path = url.getPath();
      int idx1 = path.lastIndexOf(".jar!");
      if (idx1 == -1) {
        idx1 = path.lastIndexOf(".zip!");
      }
      int idx2 = path.lastIndexOf(".jar!", idx1 - 1);
      if (idx2 == -1) {
        idx2 = path.lastIndexOf(".zip!", idx1 - 1);
      }
      if (idx2 == -1) {
        File file = new File(decodeURIComponent(path.substring(5, idx1 + 4), false));
        zip = new ZipFile(file);
      } else {
        String s = path.substring(idx2 + 6, idx1 + 4);
        File file = resolveFile(s);
        zip = new ZipFile(file);
      }

```

jar:file:/Users/.../target/wageapi-0.0.1-SNAPSHOT.war!/static

### Do you have a reproducer?

no

### Steps to reproduce

1. put FreeMarker template package in classpath e.g. /static/iframe.ftl
2. package vertx app as war (not jar or zip)
3. run package as java -jar file.war
4. call FreeMarkerTemplateEngine.create(vertx).render(data, "/static/iframe.ftl", response -> {
5. get the error (see below)

### Extra

```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 5, end 3, length 34
        at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4602)
        at java.base/java.lang.String.substring(String.java:2705)
        at io.vertx.core.file.impl.FileResolverImpl.unpackFromJarURL(FileResolverImpl.java:291)
        at io.vertx.core.file.impl.FileResolverImpl.unpackUrlResource(FileResolverImpl.java:239)
        at io.vertx.core.file.impl.FileResolverImpl.resolveFile(FileResolverImpl.java:162)
        at io.vertx.core.impl.VertxImpl.resolveFile(VertxImpl.java:813)
        at io.vertx.core.file.impl.FileSystemImpl$20.perform(FileSystemImpl.java:1135)
```

--- a/src/main/java/io/vertx/core/file/impl/FileResolverImpl.java
+++ b/src/main/java/io/vertx/core/file/impl/FileResolverImpl.java
@@ -279,24 +279,30 @@ public class FileResolverImpl implements FileResolver {
     ZipFile zip = null;
     try {
       String path = url.getPath();
-      int idx1 = path.lastIndexOf(".jar!");
-      if (idx1 == -1) {
-        idx1 = path.lastIndexOf(".zip!");
+      int idx1 = -1, idx2 = -1;
+      for (int i = path.length() - 1; i > 4; ) {
+        if (path.charAt(i) == '!' && (path.startsWith(".jar", i - 4) || path.startsWith(".zip", i - 4) || path.startsWith(".war", i - 4))) {
+          if (idx1 == -1) {
+            idx1 = i;
+            i -= 4;
+            continue;
+          } else {
+            idx2 = i;
+            break;
+          }
+        }
+        i--;
       }
-      int idx2 = path.lastIndexOf(".jar!", idx1 - 1);
       if (idx2 == -1) {
-        idx2 = path.lastIndexOf(".zip!", idx1 - 1);
-      }
-      if (idx2 == -1) {
-        File file = new File(decodeURIComponent(path.substring(5, idx1 + 4), false));
+        File file = new File(decodeURIComponent(path.substring(5, idx1), false));
         zip = new ZipFile(file);
       } else {
-        String s = path.substring(idx2 + 6, idx1 + 4);
+        String s = path.substring(idx2 + 2, idx1);
         File file = resolveFile(s);
         zip = new ZipFile(file);
       }
 
-      String inJarPath = path.substring(idx1 + 6);
+      String inJarPath = path.substring(idx1 + 2);
       StringBuilder prefixBuilder = new StringBuilder();
       int first = 0;
       int second;

--- a/src/test/java/io/vertx/core/file/WarFileResolverTest.java
+++ b/src/test/java/io/vertx/core/file/WarFileResolverTest.java
@@ -1,0 +1,47 @@
+/*
+ * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.file;
+
+import io.vertx.test.core.TestUtils;
+import org.junit.Assert;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+
+public class WarFileResolverTest extends FileResolverTestBase {
+
+  static File getFiles(File baseDir) throws Exception {
+    File file = Files.createTempFile(TestUtils.MAVEN_TARGET_DIR.toPath(), "", "files.war").toFile();
+    Assert.assertTrue(file.delete());
+    return ZipFileResolverTest.getFiles(
+      baseDir,
+      file,
+      out -> {
+        try {
+          return new JarOutputStream(out);
+        } catch (IOException e) {
+          throw new AssertionError(e);
+        }
+      }, JarEntry::new);
+  }
+
+  @Override
+  protected ClassLoader resourcesLoader(File baseDir) throws Exception {
+    File files = getFiles(baseDir);
+    return new URLClassLoader(new URL[]{files.toURI().toURL()}, Thread.currentThread().getContextClassLoader());
+  }
+}

